### ISSUE TITLE
Extend VI loss and gradients methods with analytic KL to dependent posteriors

### ISSUE THREAD

**Issue by dustinvtran** [blei-lab/edward#783]
See discussion in https://discourse.edwardlib.org/t/rookie-problem-klqp-gets-obviously-wrong-result/381/7.

**Comment by bkayalibay** [comment]
A while ago, we (@bayerj, @taimir) ran into a similar issue while working with Edward on sequential latent variable models. I'll try to outline it here to provide another case where this issue is relevant.

### The Setting
The problem occurs when all latent variables are instances of `Normal` and there are dependencies between some of the priors in our directed graphical model. This is for instance the case in sequential latent variable models with Markovian state transition and models with multiple layers of latent variables (like the ones in [1](https://arxiv.org/abs/1401.4082v3) and [2](https://arxiv.org/pdf/1509.00519v4.pdf)).

### The Problem
In the above setting, the following applies for the kl divergence over the set of all latent variables in our model (I will call this `Z`):

![kl_set](https://user-images.githubusercontent.com/8435918/30847669-a5c8da7a-a29c-11e7-9ebc-7fc377ca6ba7.gif)

The final term here suggests that we must condition each prior random variable on the ancestors of `z`, where `z` is a sample from the approximate posterior. `build_reparam_loss_and_gradients` respects this by copying each latent variable `z` and replacing all samples from the prior by those from the posterior in the computational graph of `z`:
```python
for z in six.iterkeys(inference.latent_vars):
      z_copy = copy(z, dict_swap, scope=scope)
      p_log_prob[s] += tf.reduce_sum(
inference.scale.get(z, 1.0) * z_copy.log_prob(dict_swap[z]))
```
`build_reparam_kl_loss_and_gradients`, on the other hand, does not do this. This is not exactly wrong behavior, since (as suggested in the docstring) this function assumes the kl divergence has an analytic solution, which is not satisfied in the setting I described. However, the class`KLqp`will automatically decide to use the analytic kl divergence if all latent variables are Normal, causing the wrong kl divergence to be calculated. 

One could prevent this by explicitly using the subclass `ReparameterizationKLqp` but one first needs to be aware that this difference in behavior exists to be able to do that and I don't think that is perfectly clear (for a long time I thought the only difference between the two functions was that the one computes a monte carlo estimate while the other uses the closed-form solution).

Also, even though the complete kl does not have an analytic solution, parts of the kl divergence still have analytic solutions (since each latent variable is `Normal`). One would like to be able to calculate these terms analytically, while using monte carlo sampling to do the conditioning on the ancestors. This is a sort of hybrid between the analytic solution and monte carlo estimation that will have less noisy gradients than pure monte carlo.

### A toy example
The script on the forum already provides a toy example of the issue. I wrote [another short script](https://gist.github.com/bkayalibay/58379f78bef040ec833216ba02ba4241) to show with plots that  `ReparameterizationKLKLqp` and  `ReparameterizationKLqp` behave differently. In my setting, I have two layers of latent variables `z1` and `z2`, where the prior of `z2` is defined as: `Normal(max(0, z1), 1)`. The prior and posterior for `z1` are both Normal. The former has a mean of 20, while the latter has a mean of -20. The posterior for  `z2` is a standard normal distribution.

It is easy to see that if we condition on samples from the posterior, `z2` will also be a standard normal and `kl(z2, qz2)` will be zero. If the conditioning does not occur, this divergence will be higher than zero. The script creates this scenario twice: once with `ReparameterizationKLKLqp` (which uses the analytic kl) and once with `ReparameterizationKLqp` (which uses the monte carlo estimate and conditions on the ancestors). I collect samples from the kl divergence 50 times (taking 10 monte carlo samples each time) and then plot these at the end of the script. The plots show that the two kl divergences are not the same.

### The Solution
As @dustinvtran wrote on the forum, the solution would be to take the expectation of the kl divergence over multiple samples by moving the relevant part of the code into the loop that precedes it. I feel like this might be inefficient when the latent variables are independent, as we would be calculating the same kl over and over (not sure if tensorflow would optimize this away). Perhaps one could use something like `is_independent` to detect dependencies between latent variables?

### References

[1] https://arxiv.org/abs/1401.4082v3
[2] https://arxiv.org/abs/1509.00519v4
